src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtAmazonDynamoDbBySharedTable.java

```diff
@@ -313,18 +313,32 @@ public PutItemResult putItem(PutItemRequest putItemRequest) {
      * TODO: write Javadoc.
      */
     public QueryResult query(QueryRequest queryRequest) {
-        // map table name
-        queryRequest = queryRequest.clone();
         TableMapping tableMapping = getTableMapping(queryRequest.getTableName());
-        queryRequest.withTableName(tableMapping.getPhysicalTable().getTableName());
+
+        // map table name
+        final QueryRequest clonedQueryRequest = queryRequest.clone();
+        clonedQueryRequest.withTableName(tableMapping.getPhysicalTable().getTableName());
 
         // map query request
-        tableMapping.getQueryAndScanMapper().apply(queryRequest);
+        //clonedQueryRequest.setExpressionAttributeNames(Optional.ofNullable(clonedQueryRequest.getFilterExpression())
+        //    .map(s -> new HashMap<>(clonedQueryRequest.getExpressionAttributeNames())).orElseGet(HashMap::new));
+        //clonedQueryRequest.setExpressionAttributeValues(Optional.ofNullable(clonedQueryRequest.getFilterExpression())
+        //    .map(s -> new HashMap<>(clonedQueryRequest.getExpressionAttributeValues())).orElseGet(HashMap::new));
+        tableMapping.getQueryAndScanMapper().apply(clonedQueryRequest);
+
+        // keep moving forward pages until we find at least one record for current tenant or reach end
```

Comment:
We shouldn't have to do that for queries (and I am trying to get rid of it for scans, because unbounded looping in the client thread causes issues): queries specify a hash key and all results for a given hash key are for the same tenant. I think the previous version of the code is sufficient.